LDAP improvements

Elegant996 opened the pull request:
This PR introduces two commits which improve the LDAP authentication system that mealie currently uses.

The first commit of the PR adds support for a `LDAP_BASE_DN` field which allows clients to properly specify the starting point of the directory for which to lookup clients. This **also** resolves an issue where Active Directory users would have issues authenticating as AD users use the form `{}@example.com` instead of the User DN. Both are now supported.

In addition to the above, the first commit also populates the `user` in the mealie DB with the approriate values from the LDAP server for `full_name` and `email` instead of just copying `username` for all fields. A check was also added to lookup the `user` email in AD and then authenticate against their `username` should there be a match in the mealie DB. This allows LDAP clients to login using their `email` on _subsequent_ logins.

The second commit adds support for secure LDAP connections. At present, mealie relies on non-secure LDAP but not all servers accept non-secure connections. This is achieved through two additional parameters: `LDAP_TLS_INSECURE` and `LDAP_TLS_CACERTFILE`, both of which are optional. Clients may either disable verification or specify the require certificate to establish the secure connection.

Thanks!

Elegant996 committed:
Use Base DN for LDAP and fetch user attrs

Requires that a Base DN be set for LDAP
Set `full_name` and `email` based on LDAP attributes when creating user

Elegant996 committed:
Add support for secure LDAP

Allow insecure LDAP connection (disabled by default)
Use CA when connecting to secure LDAP server

Elegant996 committed:
Added missing quotes to example

hay-kot commented:
From the discussion

> @hay-kot Reviewed the PR again, should be good to go. Not sure how to prevent the error regarding security.py from generating. Thanks!

What have you tried to fix them? 

Looks like you need to run black to fix your formatting.

hay-kot commented on mealie/core/security/security.py:
```diff
@@ -47,23 +47,38 @@ def user_from_ldap(db: AllRepositories, session, username: str, password: str) -
         return False
 
-    user = db.users.get_one(username, "username", any_case=True)
+    user_dn, user_attr = conn.search_s(settings.LDAP_BASE_DN, ldap.SCOPE_SUBTREE, '(&(objectClass=user)(|(cn=%s)(sAMAccountName=%s)(mail=%s)))' % (username, username, username), ['name', 'mail'])[0]
```
Can you provide a code comment to explain what/why this string is needed? Also, you should probably use an f string instead of an interpolated string, unless there's a specific reasons that I'm missing?

Elegant996 replied:
Done and done.

Elegant996 committed:
Update security.py

Elegant996 committed:
Update security.py formatting

Elegant996 committed:
Update security.py

Switched to f-String formatting

hay-kot committed:
formatting

hay-kot commented:
Hey, I rebased this PR with the intention of merging, but the LDAP tests are failing. 

Can you fix these and verify they're running on your machine and convert the back from a draft?

Elegant996 committed:
Merge branch 'hay-kot:mealie-next' into ldap-improv

Elegant996 committed:
Update test_security.py

Added at attributes for testing

Elegant996 commented:
Looks like we need to include the attributes when testing, should be good now...

Elegant996 committed:
Update test_security.py

Modified tests for base DN

Elegant996 committed:
Update test_security.py

Set proper base DN for testing

Elegant996 committed:
Update test_security.py

Corrected testing for LDAP

Elegant996 committed:
Update test_security.py

Defined base_dn

Elegant996 committed:
Authenticated user not in base DN

Add check for when user can authenticate but is not in base DN

Elegant996 committed:
Update test_security.py

LDAP user cannot exist as it is searched before it is created and the list returns False

Elegant996 commented:
Resolved this but had to change the check for the result to false. The old method created the user and then searched against it in LDAP to determine if it had admin capability. The new method searches against the username to determine if it exists in LDAP before creating it. Since the user does not exist (dummy LDAP), an empty list of users is returned which ends in false.
